components/tidb_query/src/rpn_expr/impl_json.rs

```diff
@@ -27,6 +30,49 @@ fn json_array(args: &[&Option<Json>]) -> Result<Option<Json>> {
     )))
 }
 
+fn json_object_validator(expr: &tipb::Expr) -> Result<()> {
+    for chunk in expr.get_children().chunks(2) {
+        if chunk.len() == 1 {
+            return Err(other_err!(
+                "Incorrect parameter count in the call to native function 'JSON_OBJECT'"
+            ));
+        } else {
+            super::function::validate_expr_return_type(&chunk[0], EvalType::Bytes)?;
+            super::function::validate_expr_return_type(&chunk[1], EvalType::Json)?;
+        }
+    }
+    Ok(())
+}
+
+/// Required args like `&[(&Option<Byte>, &Option<Json>)]`.
+#[rpn_fn(raw_varg, extra_validator = json_object_validator)]
+#[inline]
+fn json_object(raw_args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    let mut pairs = BTreeMap::new();
+    for chunk in raw_args.chunks(2) {
+        // chunk.len() must be 1 or 2 here.
+        let key = Evaluable::borrow_scalar_value_ref(&chunk[0])
+            .as_ref()
+            .map_or_else(
+                || {
+                    Err(other_err!(
+                        "Data truncation: JSON documents may not contain NULL member names."
+                    ))
+                },
+                |v: &Bytes| {
+                    String::from_utf8(v.to_owned()).map_err(|e| crate::codec::Error::from(e).into())
```

Comment:
Could we use `?` to omit `.into`

Reply:
> Could we use `?` to omit `.into`

Err here is a `FromUtf8Error`, and only `crate::codec::Error` implements `From<FromUtf8Error>`, so I use `.map_err(|e| crate::codec::Error::from(e).into())` here.

If we want to use `?`, maybe we are going to use `Ok(String::from_utf8(v.to_owned()).map_err(|e| crate::codec::Error::from(e))?)` ?